feat(git): Use local git for conflict detection

zharinov opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->

## Changes:

- Remove `isConflicted` field from platform and worker code
- Call `git.isBranchConflicted()` instead of using `isConflicted`

## Context:

Closes #3866

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

zharinov commented:
Unfortunately, this PR can't be split to separate smaller chunks

rarkins commented on lib/util/git/index.ts:
```diff
@@ -551,6 +551,30 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+  try {
+    await syncGit();
+    await git.reset(ResetMode.HARD);
+    await git.checkout(['-B', branchName, 'origin/' + branchName]);
```
Why is it needed to checkout the origin branch? Maybe document it for future

zharinov replied:
Oops, it's the result of copy-paste (fixed now)

rarkins commented on lib/util/git/index.ts:
```diff
@@ -551,6 +551,30 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+    await syncGit();
+    await git.reset(ResetMode.HARD);
+    await git.checkout(['-B', branchName, 'origin/' + branchName]);
+    await git.merge(['--no-commit', '--no-ff', baseBranch]);
```
Shouldn't there be a reset afterwards?

rarkins commented on lib/workers/branch/reuse.ts:
```diff
@@ -70,7 +75,8 @@ export async function shouldReuseExistingBranch(
-  if (pr?.isConflicted) {
+  const isConflicted = await isBranchConflicted(baseBranch, branchName);
+  if (isConflicted) {
     logger.debug('PR is conflicted');
```
```suggestion
    logger.debug('Branch is conflicted');
```

rarkins commented on lib/util/git/index.ts:
```diff
@@ -551,6 +551,30 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
   return true;
 }
 
+export async function isBranchConflicted(
```
If this command is called in both the branch worker and PR worker, we should find a way to cache the result, either in the git layer or by adding the result to the branch config that's passed to the PR worker.

JamieMagee commented on lib/util/git/index.ts:
```diff
@@ -551,6 +551,32 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+    await git.merge(['--no-commit', '--no-ff', baseBranch]);
+    await git.reset(ResetMode.HARD);
+    await git.checkout(baseBranch);
+  } catch (err) {
```
log this?

zharinov commented:
Ready for review

viceice commented on lib/util/git/index.ts:
```diff
@@ -551,6 +551,30 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+    if (err?.git?.conflicts?.length) {
+      return true;
+    }
+    logger.debug({ branchName, err }, 'Conflict detection: unknown error');
```
All other errors as no conflict?

zharinov replied:
We may return `null` for errors, but consuming part won't treat it differently. Alternatively, we may treat any error as `this brach is conflicted`, but most probably it's incorrect too 🤔

viceice commented on lib/platform/types.ts:
```diff
@@ -55,7 +55,6 @@ export interface Pr {
   displayNumber?: string;
   hasAssignees?: boolean;
   hasReviewers?: boolean;
-  isConflicted?: boolean;
```
I think we should also use this as fast conflict check if available, so we don't need the manual merge test. That will speedup checks.

rarkins replied:
Depends. For some platforms like Bitbucket it involves another API call specifically

rarkins replied:
Also, I think @zharinov found in testing that GitHub even returns UNKNOWN instead of CONFLICTED many times, so it's not reliable (we know we can't merge it, but we may miss that it's conflicted and not rebase it)

viceice replied:
sure, but some platforms return conflicted as pr staste, so we don't need the git check. So i would prefer to leave the pr check when it's part of the fetched pr metadata and only use git check if pr is not already conflicted

rarkins replied:
GitHub is not reliable (it can return UNKNOWN if it hasn't calculated CONFLICTED status yet), Bitbucket takes API call. Not sure it's worth then the special case of using PR.isConflicted if it's only for a subset. Also would be interested to learn how long this command takes to execute locally - @zharinov can you do some timing locally?

rarkins replied:
BTW I suspect that the reason we sometimes see conflicted PRs not rebased when we expect to is because the GitHub backend is determining conflicted state asynchronously and returning UNKNOWN

zharinov replied:
> From what I see, for non-conflicted branches it's about 100ms, while for the conflicted one it's about 1100ms

zharinov replied:
UPD: same results for measuring started after `syncGit()` call

rarkins replied:
> > can you do some timing locally?
> 
> From what I see, for non-conflicted branches it's about 100ms, while for the conflicted one it's about 1100ms

Can you further work out which of the steps takes up that time? e.g. is it mostly the attempted merge, or is it partly the branch checkouts and resets, etc?

Also I think we should aim to cache this information. e.g. record the baseBranchSha and branchSha and conflicted result and then reuse it in subsequent runs if unchanged - meaning only need to recompute it whenever the base branch is updated.

zharinov replied:
Sure, I'll research it for more details

zharinov replied:
> Can you further work out which of the steps takes up that time? e.g. is it mostly the attempted merge, or is it partly the branch checkouts and resets, etc?

As I can see, it's mostly because of syncing with remote. When we do checkout to `renovate/some-branch` for the first time manually, it shows me progress info on how remote objects are being received. And if I get it right, we're about to have this lag anyway, attempting to rebase branch. Other steps seem to be relatively instant (~20ms each).

But we have additional 500ms for each conflict check, and this is subject to my further work on caching.

rarkins replied:
Yes, we shouldn't count the sync if we will need to do it anyway. I'm interested in the time to calculate the merge and reset, both in conflicted and non-conflicted state

zharinov commented on lib/util/git/index.ts:
```diff
@@ -24,6 +24,7 @@ import { logger } from '../../logger';
 import { ExternalHostError } from '../../types/errors/external-host-error';
 import { GitOptions, GitProtocol } from '../../types/git';
 import { Limit, incLimitedValue } from '../../workers/global/limits';
+import * as packageCache from '../cache/package';
```
Can we use this type of cache here?

rarkins replied:
No

zharinov replied:
Then, module-local cache inside `config` variable?

rarkins replied:
It's likely to be unique due to the SHAs.. but it should instead be part of the repository cache. We should keep indefinitely a list of all base -> branch conflicted values we still need and have a routine to clean out old ones which are no longer needed.

zharinov commented on lib/util/git/index.ts:
```diff
@@ -551,6 +552,69 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+
+  const cache = getCache();
+  const olderBranchSha = cache.baseBranchSha;
+  if (olderBranchSha !== baseBranchSha) {
```
Any change of base branch leads to cache reset

rarkins commented on lib/util/cache/repository/types.ts:
```diff
@@ -40,4 +40,6 @@ export interface Cache {
   init?: RepoInitConfig;
   scan?: Record<string, BaseBranchCache>;
   lastPlatformAutomergeFailure?: string;
+  baseBranchSha?: string;
```
I don't think this is the correct place. There's no single `baseBranchSha` in a repo. There's not only multiple base branch names.. but we will have multiple PRs merged

rarkins commented:
Maybe as a first step we do it *without* caching?

zharinov commented:
It seems I was wrong when told that it takes 1s for dirty branches and near instant for clean ones. What I noticed later, it takes at least 500ms for *any* branch because of remote syncing. Seems pretty costly to me, isn't it?

rarkins commented:
> It seems I was wrong when told that it takes 1s for dirty branches and near instant for clean ones. What I noticed later, it takes at least 500ms for _any_ branch because of remote syncing. Seems pretty costly to me, isn't it?

That is, but I wonder why remote syncing is necessary. Don't we fully sync when we clone? And if so then why do we need to reference an `origin/` branch here?

rarkins commented:
If the origin repo receives a commit to a base branch *after* we clone, it's not our intention that we detect it and keep going. If we don't detect it then it's fine. If somehow it causes a problem (e.g. when we push a branch) then we abort with repository-changed error. I thought for isConflicted we could do our calculation entirely locally.

zharinov commented:
> That is, but I wonder why remote syncing is necessary. Don't we fully sync when we clone? And if so then why do we need to reference an origin/ branch here?

It has something to do with `--filter=blob:none` option.

Clone with blobs:
```sh
$ git clone git@github.com:renovate-testing/foobar.git /tmp/foo
$ cd /tmp/foo
$ git checkout bar
Branch 'bar' set up to track remote branch 'bar' from 'origin'.
Switched to a new branch 'bar'
```

Clone without blobs:
```sh
$ git clone git@github.com:renovate-testing/foobar.git /tmp/bar --filter=blob:none
$ cd /tmp/bar
$ git checkout bar
remote: Enumerating objects: 1, done.
remote: Counting objects: 100% (1/1), done.
remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (1/1), 45 bytes | 45.00 KiB/s, done.
Branch 'bar' set up to track remote branch 'bar' from 'origin'.
Switched to a new branch 'bar'
```

rarkins commented:
Experimenting with Renovate's repo after a blobless clone:

```
❯ git branch --all --format='%(refname:short)'
main
origin/HEAD
origin/ci/test-node-v16
origin/docs/docusaurus
origin/feat/10135-docker-registry-url
origin/feat/4404-rollup
origin/feat/6508-julia
origin/feat/docker-registry-url
origin/feat/lowercase-proxy-env-vars
origin/feat/npm-install-tool
origin/feat/standard-version-changelog-support
origin/feat/supress-lfs-pull
origin/fix/11944-goproxy-fallback
origin/fix/gradle-qualifier
origin/fix/revert-11565
origin/fix/terraform-server-error
origin/main
origin/refactor/manager-types
origin/refactor/packagist-datasource
origin/refactor/url-module
origin/renovate/actions-setup-java-2.x
origin/renovate/actions-setup-node-2.x
origin/v26
origin/viceice-patch-1

❯ git branch --format='%(refname:short)'
main
```

rarkins commented:
This is for a branch of mine which is 6 commits ahead, 144 commits behind `main`:

```
❯ time git merge origin/feat/npm-install-tool --no-commit --no-ff
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 10 (delta 6), reused 0 (delta 0), pack-reused 1
Receiving objects: 100% (10/10), 7.92 KiB | 7.92 MiB/s, done.
Resolving deltas: 100% (6/6), done.
Auto-merging lib/manager/npm/post-update/lerna.ts
Auto-merging lib/manager/npm/post-update/npm.ts
Auto-merging lib/manager/npm/post-update/pnpm.ts
Auto-merging lib/manager/npm/post-update/yarn.ts
Auto-merging lib/util/exec/buildpack.ts
CONFLICT (content): Merge conflict in lib/util/exec/buildpack.ts
Automatic merge failed; fix conflicts and then commit the result.
git merge origin/feat/npm-install-tool --no-commit --no-ff  0.11s user 0.08s system 28% cpu 0.665 total
```

rarkins commented:
This one is 326 commits behind main:

```
❯ time git merge origin/feat/10135-docker-registry-url --no-commit --no-ff
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 11 (delta 3), reused 0 (delta 0), pack-reused 2
Receiving objects: 100% (11/11), 11.28 KiB | 11.28 MiB/s, done.
Resolving deltas: 100% (3/3), done.
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 4 (delta 2), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (4/4), 1.51 KiB | 1.51 MiB/s, done.
Resolving deltas: 100% (2/2), done.
Auto-merging lib/datasource/docker/readme.md
CONFLICT (add/add): Merge conflict in lib/datasource/docker/readme.md
Auto-merging lib/manager/dockerfile/extract.spec.ts
CONFLICT (content): Merge conflict in lib/manager/dockerfile/extract.spec.ts
Auto-merging lib/manager/dockerfile/extract.ts
CONFLICT (content): Merge conflict in lib/manager/dockerfile/extract.ts
Auto-merging lib/manager/kustomize/__snapshots__/extract.spec.ts.snap
CONFLICT (content): Merge conflict in lib/manager/kustomize/__snapshots__/extract.spec.ts.snap
Auto-merging lib/manager/kustomize/extract.spec.ts
Auto-merging lib/manager/terraform/__snapshots__/extract.spec.ts.snap
Automatic merge failed; fix conflicts and then commit the result.
git merge origin/feat/10135-docker-registry-url --no-commit --no-ff  0.21s user 0.13s system 20% cpu 1.716 total
```

rarkins commented:
@zharinov do you see any barrier to checking out the base branch and then testing merging from the renovate branch into it? I don't know if it's necessarily faster, but I think it's cleaner to just stay in e.g. `main` and then test the others merging into it.

zharinov commented:
> do you see any barrier to checking out the base branch and then testing merging from the renovate branch into it?

True, it's one of the options. I tried it and it worked. But, as you said, there is no performance advantage.

rarkins commented:
Here's the same examples where we first checkout the branch and then test the merge:

```
❯ time bash -c "git checkout feat/npm-install-tool && git merge main --no-commit --no-ff"
remote: Enumerating objects: 270, done.
remote: Counting objects: 100% (238/238), done.
remote: Compressing objects: 100% (233/233), done.
remote: Total 270 (delta 17), reused 5 (delta 5), pack-reused 32
Receiving objects: 100% (270/270), 659.08 KiB | 6.87 MiB/s, done.
Resolving deltas: 100% (22/22), done.
Branch 'feat/npm-install-tool' set up to track remote branch 'feat/npm-install-tool' from 'origin'.
Switched to a new branch 'feat/npm-install-tool'
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 5 (delta 2), reused 0 (delta 0), pack-reused 1
Receiving objects: 100% (5/5), 6.55 KiB | 6.55 MiB/s, done.
Resolving deltas: 100% (2/2), done.
Auto-merging lib/manager/npm/post-update/lerna.ts
Auto-merging lib/manager/npm/post-update/npm.ts
Auto-merging lib/manager/npm/post-update/pnpm.ts
Auto-merging lib/manager/npm/post-update/yarn.ts
Auto-merging lib/util/exec/buildpack.ts
CONFLICT (content): Merge conflict in lib/util/exec/buildpack.ts
Automatic merge failed; fix conflicts and then commit the result.
bash -c   0.33s user 0.31s system 36% cpu 1.752 total

/tmp/script/renovate feat/npm-install-tool|merge*
❯ git reset --hard
HEAD is now at 68253c9cb Update lib/manager/npm/post-update/lerna.ts

/tmp/script/renovate feat/npm-install-tool
❯ time bash -c "git checkout feat/10135-docker-registry-url && git merge main --no-commit --no-ff"
remote: Enumerating objects: 497, done.
remote: Counting objects: 100% (445/445), done.
remote: Compressing objects: 100% (431/431), done.
remote: Total 497 (delta 50), reused 14 (delta 14), pack-reused 52
Receiving objects: 100% (497/497), 886.10 KiB | 9.04 MiB/s, done.
Resolving deltas: 100% (50/50), done.
Updating files: 100% (544/544), done.
Branch 'feat/10135-docker-registry-url' set up to track remote branch 'feat/10135-docker-registry-url' from 'origin'.
Switched to a new branch 'feat/10135-docker-registry-url'
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 5 (delta 0), reused 0 (delta 0), pack-reused 1
Receiving objects: 100% (5/5), 7.30 KiB | 7.30 MiB/s, done.
Auto-merging lib/datasource/docker/readme.md
CONFLICT (add/add): Merge conflict in lib/datasource/docker/readme.md
Auto-merging lib/manager/dockerfile/extract.spec.ts
CONFLICT (content): Merge conflict in lib/manager/dockerfile/extract.spec.ts
Auto-merging lib/manager/dockerfile/extract.ts
CONFLICT (content): Merge conflict in lib/manager/dockerfile/extract.ts
Auto-merging lib/manager/kustomize/__snapshots__/extract.spec.ts.snap
CONFLICT (content): Merge conflict in lib/manager/kustomize/__snapshots__/extract.spec.ts.snap
Auto-merging lib/manager/kustomize/extract.spec.ts
Auto-merging lib/manager/terraform/__snapshots__/extract.spec.ts.snap
Automatic merge failed; fix conflicts and then commit the result.
bash -c   0.44s user 0.52s system 42% cpu 2.261 total
```

Based on that non-scientific example, it's 2-3x slower than starting on `main`. I don't think there's any need to test further, but it reenforces the decision to start on `main` and merge into it.

rarkins commented on lib/util/git/index.ts:
```diff
@@ -568,6 +568,43 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+        'Conflict detection: unknown error'
+      );
+    }
+  }
+
+  return result;
```
I think it needs the `git merge --abort` added here (within its own try/catch maybe?

zharinov replied:
Ah, now I see we can't rely on `syncGit()` for cleaning it up

rarkins commented on lib/workers/branch/reuse.ts:
```diff
@@ -58,10 +65,11 @@ export async function shouldReuseExistingBranch(
       if (await isBranchModified(branchName)) {
         // TODO: Warn here so that it appears in PR body (#9720)
         logger.debug('Cannot rebase branch as it has been modified');
-        return { reuseExistingBranch: true, isModified: true };
+        result.reuseExistingBranch = true;
+        result.isModified = true;
```
Is the `return` missing here accidentally?

zharinov replied:
Yup 😅

zharinov replied:
Not exactly, though, my intent was to return couple lines after. What I missed is excessive debug statement.

rarkins reviewed (changes requested):
Can the branch reuse refactoring be separated to a different PR?

rarkins reviewed (dismissed):
I think this LGTM, although it's a pretty big change so we need to keep our eyes open for performance or functional problems afterwards

zharinov commented:
Stay tuned, I'll double check this locally with all platforms

zharinov commented:
> Stay tuned, I'll double check this locally with all platforms

Well, I can confirm everything worked with Azure, GitLab and GitHub. Most probaby it's okay for other platforms as well :shipit:

viceice commented:
Ok, will review tomorrow

viceice commented on lib/platform/bitbucket/index.ts:
```diff
@@ -1,7 +1,6 @@
 import URL from 'url';
 import is from '@sindresorhus/is';
 import JSON5 from 'json5';
-import parseDiff from 'parse-diff';
```
Is this dep now obsolete and can be removed?

viceice commented on lib/util/git/index.spec.ts:
```diff
@@ -600,4 +600,90 @@ describe('util/git/index', () => {
+    });
+
+    afterAll(async () => {
+      const repo = Git(base.path);
```
Why do we need this? Isn't the repo deleted anyways after all tests?

viceice reviewed (commented):
We should add caching very soon after this.

zharinov commented:
> We should add caching very soon after this.

Already working on this
